### ISSUE TITLE
Proposal: Add Error JobConditionType to reflect controller error (Resource Quota Error) into the status

### ISSUE THREAD

**Issue by CJZheng91** [kubeflow/common#47]
**Problem**: Currently JobConditionType constants, there doesn't exist a a status that indicates error happening in the controller. This cause problems, e.g. when pod creation fails because it exceeds resource quota. Take TFJob as an example (https://github.com/kubeflow/tf-operator/blob/5adee6f30c86484897db33188af591d5976d1cd2/pkg/control/pod_control.go#L138), if pod creation here exceeds resource quota, the `Create` func will return an error. This information is only exposed to the higher-level jobs as an event. 

**Solution**: Add an Error JobConditionType in kubeflow common api. Also show pod creation error (or other types of error if necessary) into the JobStatus, e.g. in the `updateStatusSingle` func for TFJob (https://github.com/kubeflow/tf-operator/blob/5adee6f30c86484897db33188af591d5976d1cd2/pkg/controller.v1/tensorflow/status.go#L61). This Error status will just indicate a retriable error that happened in the controller.

**Comment by czheng94** [comment]
Any thoughts on this issue?

Some updates supporting the necessity of the Error ConditionType:

Similar "Error" conditions do exist in other Kubernetes components, e.g. in Deployment and ReplicaSet. 

`ReplicaSet` has `ReplicaSetReplicaFailure` as a condition type:
https://github.com/kubernetes/kubernetes/blob/abe6321296123aaba8e83978f7d17951ab1b64fd/pkg/apis/apps/types.go#L788-L793

`Deployment` has `DeploymentReplicaFailure` as a condition type:
https://github.com/kubernetes/kubernetes/blob/abe6321296123aaba8e83978f7d17951ab1b64fd/pkg/apis/apps/types.go#L460-L473

`Deployment` controller will check for underlying `ReplicaSet` condition type and propagate `ReplicaSetReplicaFailure` up into `Deployment` status:
https://github.com/kubernetes/kubernetes/blob/abe6321296123aaba8e83978f7d17951ab1b64fd/pkg/controller/deployment/progress.go#L101-L106


Right now, pod creation errors due to resource quota are only exposed as events in KubeFlow jobs, which is not a very good channel to expose and especially to propagate errors. Enabling this Error condition in KubeFlow job CR status will enable much easier error propagation for users who are building custom CR on top of KubeFlow jobs.

**Comment by terrytangyuan** [comment]
cc @Jeffwan @gaocegege @johnugeorge @ywskycn @merlintang WDYT? Any objections on bringing this type of errors to Kubeflow CR level? It would be convenient to surface this at CR level status so users don't have to check each individual subresources but my main concern is that it's hard to define the types of errors that's retriable because this depends on the specific controller, use cases, as well as the lower-level mechanism to handle/propagate errors in different clusters.

**Comment by czheng94** [comment]
@terrytangyuan 
> but my main concern is that it's hard to define the types of errors that's retriable because this depends on the specific controller,

I don't think we need to define these specific error types in kubeflow/common. They can just all be the "Error" JobConditionType, then in each specific controller, we can use `JobCondition.Reason` and `JobCondition.Message` to expose the specific error.
